### PR TITLE
feat: adds tx result to network substates

### PR DESCRIPTION
**Pull request by jorgeantonio21**
Description
---
Adds a `TransactionReceipt` as a substate to the network, for every successfully finalized transaction. The structure is as follows:

```rs
pub struct TransactionReceipt {
  pub transaction_hash: Hash,
  pub events: Vec<Event>,
  pub logs: Vec<LogEntry>,
  pub fee_receipt: FeeReceipt,
}
```
On every transaction submission, after the transaction has been accepted by the network, each involved substate is stored in the network ledger. Additionally, we are now storing an additional substate, which contains metadata about the transaction.

Notice that, at the moment, we decided to leave the involved `SubstateDiff` from the definition of `TransactionReceipt` as it would involve recursive calls on substates to be persisted, on the tracker storage.

Motivation and Context
---
The current goal is to index events from substates of the network.

How Has This Been Tested?
---
Checked count of substates in tests.

What process can a PR reviewer use to test or verify this change?
---

Breaking Changes
---

- [x] None
- [ ] Requires data directory to be deleted
- [ ] Other - Please specify

### CODE REVIEW

File: applications/tari_dan_wallet_cli/src/command/transaction.rs [hunk 1]
```diff
@@ -566,6 +566,9 @@ pub fn summarize_finalize_result(finalize: &FinalizeResult) {
                     SubstateValue::Resource(_) => {
                         println!("      ▶ resource: {}", address);
                     },
+                    SubstateValue::TransactionReceipt(_) => {
+                        println!("      ▶ execute_result: {}", address);
```
Comment (mrnaveira) [inline applications/tari_dan_wallet_cli/src/command/transaction.rs]: `println!("      ▶ transaction_receipt: {}", address);`

File: applications/tari_validator_node/tests/utils/validator_node_cli.rs [hunk 2]
```diff
@@ -210,6 +210,13 @@ pub(crate) fn add_substate_addresses(world: &mut TariWorld, outputs_name: String
                 });
                 counters[5] += 1;
             },
+            SubstateAddress::TransactionReceipt(_) => {
+                outputs.insert(format!("execute_results/{}", counters[6]), VersionedSubstateAddress {
```
Comment (mrnaveira) [inline applications/tari_validator_node/tests/utils/validator_node_cli.rs]: `outputs.insert(format!("transaction_receipts/{}...`

File: applications/tari_validator_node/tests/utils/wallet_daemon_cli.rs [hunk 3]
```diff
@@ -690,13 +690,20 @@ fn add_substate_addresses_from_wallet_daemon(world: &mut TariWorld, outputs_name
...
             },
             SubstateAddress::NonFungibleIndex(_) => {
                 outputs.insert(format!("nft_indexes/{}", counters[5]), VersionedSubstateAddress {
                     address: addr.clone(),
                     version: data.version(),
                 });
+                counters[5] += 1;
+            },
+            SubstateAddress::TransactionReceipt(_) => {
+                outputs.insert(format!("execute_results/{}", counters[6]), VersionedSubstateAddress {
```
Comment (mrnaveira) [inline applications/tari_validator_node/tests/utils/wallet_daemon_cli.rs]: `outputs.insert(format!("transaction_receipts/{}...`

File: applications/tari_validator_node_cli/src/command/transaction.rs [hunk 4]
```diff
@@ -403,6 +403,9 @@ fn summarize_finalize_result(finalize: &FinalizeResult) {
                     SubstateValue::Resource(_) => {
                         println!("      ▶ resource: {}", address);
                     },
+                    SubstateValue::TransactionReceipt(_) => {
+                        println!("      ▶ execute_result: {}", address);
```
Comment (mrnaveira) [inline applications/tari_validator_node_cli/src/command/transaction.rs]: `println!("      ▶ transaction_receipt: {}", address);`

File: dan_layer/engine/src/runtime/error.rs [hunk 5]
```diff
@@ -122,6 +126,10 @@ pub enum RuntimeError {
     ComponentAddressMustBeSequential { index: u32 },
     #[error("Failed to load template '{address}': {details}")]
     FailedToLoadTemplate { address: TemplateAddress, details: String },
+    #[error("Execute Result already exists {address}")]
+    TransactionReceiptAlreadyExists { address: TransactionReceiptAddress },
+    #[error("Execute Result does not exist")]
+    ExecuteResultDoesNotExist,
```
Comment (mrnaveira) [inline dan_layer/engine/src/runtime/error.rs]: I will use "Transaction Receipt" also in the error variants and descriptions

File: dan_layer/engine_types/src/commit_result.rs [hunk 6]
```diff
@@ -32,6 +35,46 @@ use crate::{
...
+
+impl<T: Into<Hash>> From<T> for TransactionReceiptAddress {
+    fn from(address: T) -> Self {
+        Self::new(address.into())
+    }
+}
+
+impl Display for TransactionReceiptAddress {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "resource_{}", self.0 .0)
```
Comment (mrnaveira) [inline dan_layer/engine_types/src/commit_result.rs]: `write!(f, "receipt_{}", self.0 .0)`

File: dan_layer/engine_types/src/indexed_value.rs [hunk 7]
```diff
@@ -20,6 +20,7 @@ pub struct IndexedValue {
     component_addresses: Vec<ComponentAddress>,
     #[serde(with = "serde_with::hex::vec")]
     resource_addresses: Vec<ResourceAddress>,
+    execute_result_addresses: Vec<TransactionReceiptAddress>,
```
Comment (mrnaveira) [inline dan_layer/engine_types/src/indexed_value.rs]: `transaction_receipt_addresses`

File: dan_layer/engine_types/src/indexed_value.rs [hunk 8]
```diff
@@ -37,6 +38,7 @@ impl IndexedValue {
             buckets: visitor.buckets,
             resource_addresses: visitor.resource_addresses,
             component_addresses: visitor.component_addresses,
+            execute_result_addresses: visitor.execute_result_addresses,
```
Comment (mrnaveira) [inline dan_layer/engine_types/src/indexed_value.rs]: `transaction_receipt_addresses`

**Review by mrnaveira — CHANGES_REQUESTED** [review]
Nice work.

I suggest:
- Fix all naming inconsistencies in the PR, use "transaction receipt" instead of "execute result" everywhere. I pointed some of them but there are more
- Add a parsing match arm for the new tx receipts addreses on `substate.rs` (`impl FromStr for SubstateAddress`). Github does not let me comment that code.

File: applications/tari_dan_wallet_cli/src/command/transaction.rs (hunk 1 above)
Comment (jorgeantonio21) [reply applications/tari_dan_wallet_cli/src/command/transaction.rs]: This is very valuable ! Thanks a lot for catching these up!

**Review by mrnaveira — DISMISSED** [review]

File: dan_layer/engine_types/src/substate.rs [hunk 9]
```diff
@@ -406,6 +421,13 @@ impl SubstateValue {
             _ => None,
         }
     }
+
+    pub fn into_transaction_receipt(&self) -> Option<&TransactionReceipt> {
```
Comment (sdbondi) [inline dan_layer/engine_types/src/substate.rs]: This is against rust convention. 

https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv

```suggestion
    pub fn into_transaction_receipt(self) -> Option<TransactionReceipt> {
```
and/or (as needed)

```suggestion
    pub fn as_transaction_receipt(&self) -> Option<&TransactionReceipt> {
```

and/or (as needed)

```suggestion
    pub fn as_transaction_receipt_mut(&mut self) -> Option<&mut TransactionReceipt> {
```

File: dan_layer/engine/src/runtime/tracker.rs [hunk 10]
```diff
@@ -667,6 +689,23 @@ impl<TTemplateProvider: TemplateProvider<Template = LoadedTemplate>> StateTracke
                 up_states.insert(addr, substate.into());
             }
 
+            if state.transaction_receipt.is_none() {
```
Comment (sdbondi) [inline dan_layer/engine/src/runtime/tracker.rs]: state.transaction_reciept is never assigned (?) I think we always want to add the transaction receipt to the substates to persist (can remove this conditional)

File: dan_layer/engine/src/runtime/tracker.rs [hunk 11]
```diff
@@ -502,16 +502,38 @@ impl<TTemplateProvider: TemplateProvider<Template = LoadedTemplate>> StateTracke
...
+                x.into_transaction_receipt().and_then(|x| {
+                    Some(TransactionReceipt {
+                        transaction_hash: x.transaction_hash,
+                        events: x.events.clone(),
+                        logs: x.logs.clone(),
+                        fee_receipt: Some(finalized_fee.clone()),
+                    })
+                });
+            })
+            .or_insert_with(|| panic!("Failed to get transaction receipt"));
```
Comment (sdbondi) [inline dan_layer/engine/src/runtime/tracker.rs]: This isnt doing anything. `and_modify` passes in a mutable reference `FnOnce(&mut V)`which you should mutate. This is mapping reference to a new TransactionReceipt and then discarding it (semi-colon at the end of the closure).

I suggest something like this
```suggestion
        let transaction_receipt_mut = substates_to_persist
            .get_mut(&SubstateAddress::TransactionReceipt(self.transaction_hash().into()))
            .and_then(|s| s.as_transaction_receipt_mut())
            .expect("invariant violation: expected transaction receipt to be added to substates to persist but it wasnt");
            
        transaction_receipt_mut.fee_receipt = Some(finalize_fee.clone());
            let events = transaction_receipt_mut
            .events
            .clone();
        let logs = transaction_receipt_mut
            .logs
            .clone();
```

File: dan_layer/engine/src/runtime/tracker.rs [hunk 12]
```diff
@@ -502,16 +502,38 @@ impl<TTemplateProvider: TemplateProvider<Template = LoadedTemplate>> StateTracke
...
+            .logs
+            .clone();
 
         // Finalise will always reset the state
         let state = self.take_working_state();
 
         let result = state
             .validate_finalized()
-            .and_then(|_| self.generate_substate_diff(state, substates_to_persist));
+            .and_then(|_| self.generate_substate_diff(state, substates_to_persist.clone()));
```
Comment (sdbondi) [inline dan_layer/engine/src/runtime/tracker.rs]: This clone should not be needed (?)
```suggestion
            .and_then(|_| self.generate_substate_diff(state, substates_to_persist));
```

File: dan_layer/engine/src/runtime/tracker.rs (hunk 11 above)
Comment (sdbondi) [reply dan_layer/engine/src/runtime/tracker.rs]: In `impl SubstateValue`
```rust
pub fn as_transaction_receipt_mut(&mut self) -> Option<&mut TransactionReceipt> {
 /// ...
}

**Review by sdbondi — COMMENTED** [review]
Nice LGTM

There is a bug: the fee receipt is not added to the substate.

**Review by sdbondi — DISMISSED** [review]

**Review by stringhandler — DISMISSED** [review]
utACK

Testing locally now

**Comment by stringhandler** [comment]
tested locally, works with one committee, but seems to fail with 2. could be unrelated

**Pull request merged**